### PR TITLE
feat: implement importing excel (.xls .xlsx) files

### LINKED ISSUE #2134: [Feature Request]: Open Excel (.xls/.xlsx) data files
### Description

Currently, only able to open .CSV files

### Purpose

Avoid having to export data files from .XLSX to .CSV, which removes all formulas, formatting, etc.

### Use-case

Opening .xlsx files directly, like JAMOVI does, saves time and reduces errors

### Is your feature request related to a problem?

Often, we do a lot of data processing in Excel (computing variables etc.). That file then has to be exported to CSV to be opened in JASP. Then, any time we need to make a change, we have to update the Excel file, re-export, then update the JASP file.

### Is your feature request related to a JASP module?

_No response_

### Describe the solution you would like

ability to directly open .xlsx Excel files

### Describe alternatives that you have considered

None

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by shun2wang**
Close: https://github.com/jasp-stats/jasp-issues/issues/2134
Close: https://github.com/jasp-stats/INTERNAL-jasp/issues/1390
Close: https://github.com/jasp-stats/INTERNAL-jasp/issues/412

JASP can now import excel files from first workbook and the first sheet by default.

I introduced a third-party library [freexl](https://www.gaia-gis.it/fossil/freexl/index) to parse the excel files. but a known limitation is that you cannot import files with more than 705 columns (this is a bug of freexl), I made a patch to lifte this limitation temporarily.

- [x] init implement import excel files (on Linux)
- [x] work for Windows
- [x] work for macOS (someone could help?)
- [x] update development document

### CODE REVIEW

**Comment by boutinb** [comment]
What nice!!! I'll try it tomorrow on Mac.

**Comment by boutinb** [comment]
I wonder why you did not use conan to load the freexl library.
I have done this for Mac (see my last commit), and it works perfect. we don't need to copy the freexl.h file in this way.
Can you try this on Windows and Linux?

File: conanfile.txt
Comment (shun2wang) [inline conanfile.txt:22]: official FreeXL has a bug, so we using a patch version from my fork.

**Comment by shun2wang** [comment]
> I wonder why you did not use conan to load the freexl library. I have done this for Mac (see my last commit), and it works perfect. we don't need to copy the freexl.h file in this way. Can you try this on Windows and Linux?

because it cannot read more than 700 cols Excel files so I made a patch and build it manually. see my first comment.

see: https://github.com/shun2wang/freexl

**Comment by shun2wang** [comment]
>see: https://github.com/shun2wang/freexl

Do we maintain a repo of JASP, I am just using my fork for now. @boutinb

**Comment by boutinb** [comment]
But if I see your PR, you are using a freeXL version set in rtools.
We could use your freeXL version on GitHub and compile it, but for maintenance reason (JASP depends on many libraries, and its build is becoming quite complex) I would prefer to lie on conan which takes care on portability and compiling issues.
That the current freeXL cannot load more than 700 columns is not a big issue I find. I would ask freeXL to merge your fix (maybe you did it already), so that the next version will include your fix.

**Comment by shun2wang** [comment]
>But if I see your PR, you are using a freeXL version set in rtools.

Yes, we using rtools msys2 build toolchain to build it on Windows and the `ReadStat` also build manually by this way. maybe @JorisGoosen know more details here.

>That the current freeXL cannot load more than 700 columns is not a big issue I find. I would ask freeXL to merge your fix (maybe you did it already), so that the next version will include your fix.

We have received complaints from users about multi-column importing, and Rens has applied a fix temporarily for that. Yes I had filed bugs to that project but I dont think it will be merged for quite some time, because freeXL is updated very infrequently to ensure stability, and the last release was a year ago.

Your concerns are valid, ~~I would try to find a way to apply patches manually from conan instead of waiting for the official master branch release.~~

EDIT: no, I used a dev branch of FreeXL which including some bugfix important. I think maybe build it manually is best choice for now.

**Comment by JorisGoosen** [comment]
I agree with Bruno that adding more manual stuff to the build is not very maintainable...
And Im already unhappy with the fact that for some systems we need to manually build readstat on some systems.

> > But if I see your PR, you are using a freeXL version set in rtools.
> 
> Yes, we using rtools msys2 build toolchain to build it on Windows and the `ReadStat` also build manually by this way. maybe @JorisGoosen know more details here.
> 
> > That the current freeXL cannot load more than 700 columns is not a big issue I find. I would ask freeXL to merge your fix (maybe you did it already), so that the next version will include your fix.

I agree that staying under 700 columns until they merge your fix is not at all a problem.
Just making it possible to import from xlsx files etc is going to be a highly appluaded feature in any case.
So very nice you made it.

> We have received complaints from users about multi-column importing, and Rens has applied a fix temporarily for that. Yes I had filed bugs to that project but I dont think it will be merged for quite some time, because freeXL is updated very infrequently to ensure stability, and the last release was a year ago.
>
> Your concerns are valid, ~I would try to find a way to apply patches manually from conan instead of waiting for the official master branch release.~
> 
> EDIT: no, I used a dev branch of FreeXL which including some bugfix important. I think maybe build it manually is best choice for now.

What bug was fixed?
Maybe it is ok to just use the version with bug that is in conan?

I will now do an indepth review of your code ^^

**Comment by JorisGoosen** [comment]
Also, if we *are* going to use our modified version I would like to move that code to jasp-stats and use that one.

**Comment by JorisGoosen** [comment]
Already a thing: in qml and c++ files we use tabs not spaces. So perhaps you could clean up your indentation to be tabs of 4 instead of the spaces?

Edit: I see that you do use tabs in a lot of places, just not everywhere?

File: Desktop/data/importers/excel/__freexl.h
Comment (JorisGoosen) [inline Desktop/data/importers/excel/__freexl.h]: Cant you just add freexl to https://github.com/jasp-stats/jasp-desktop/blob/3d01a0986f4485a19874a16426174ecfcb7342a6/Desktop/CMakeLists.txt#L92 and then do a normal include?

File: Desktop/data/importers/excel/excel.cpp
Comment (JorisGoosen) [inline Desktop/data/importers/excel/excel.cpp]: You could also use `QFileInfo` for this.

File: Desktop/data/importers/excel/excel.cpp
Comment (JorisGoosen) [inline Desktop/data/importers/excel/excel.cpp]:
```suggestion
	QString xlsFilePath = tq(_path);
	const char* utf8Path = _path.c_str(); //But it would be better to just use _path.c_str() directly if you need it. It is in utf8 in any case.
	QString extension = QFileInfo(xlsFilePath).suffix().toLower();
```

File: Desktop/data/importers/excel/excel.h
Comment (JorisGoosen) [inline Desktop/data/importers/excel/excel.h:24]: So here you include "freexl.h" from the include paths, but why then add `__freexl.h' ?

File: Desktop/data/importers/excel/excelimportcolumn.cpp
Comment (JorisGoosen) [inline Desktop/data/importers/excel/excelimportcolumn.cpp]: This not really necessary, as the vector will get cleaned up in any case by the destructor.

File: Desktop/data/importers/excelimporter.cpp
Comment (JorisGoosen) [inline Desktop/data/importers/excelimporter.cpp]: Im not a big fan of `using namespace std;` we still have it in some older code.
But adding `std::` in front of functions is not so bad and it avoids possible misdirections for functions like `max()` or `min()` and other, perhaps unexpected, collisions.

And I see that `std::` is being used anyway so it shouldnt be a problem to remove it anyway.

File: Desktop/data/importers/excelimporter.cpp
Comment (JorisGoosen) [inline Desktop/data/importers/excelimporter.cpp]:
```suggestion
				string colName = colNames[i];
				if (colName.empty()) 
					colName = "V" + std::to_string(i + 1);
				else if(ColumnUtils::isIntValue(colName))
					colName = "V" + colName;

				//What is the purpose of the `find` here? Maybe explain in a comment here
				if (std::find(colNames.begin(), colNames.begin() + i, colName) != colNames.begin() + i)
					colName += "_" + std::to_string(i + 1);
```

File: Desktop/data/importers/excelimporter.cpp
Comment (JorisGoosen) [inline Desktop/data/importers/excelimporter.cpp]:
```suggestion
		else 
			for (int i = 0; i < importColumns.size(); ++i) 
				importColumns[i]->addValue(i < lineValues.size() ? lineValues[i] : "");
```

File: Desktop/data/importers/excelimporter.cpp
Comment (JorisGoosen) [inline Desktop/data/importers/excelimporter.cpp]:
```suggestion
	for (ExcelImportColumn* col : importColumns) 
		data->addColumn(col);
```

File: Desktop/data/importers/excelimporter.cpp
Comment (JorisGoosen) [inline Desktop/data/importers/excelimporter.cpp:25]:
```suggestion
#include "columnutils.h"

```

File: Desktop/data/importers/excel/__freexl.h
Comment (boutinb) [reply Desktop/data/importers/excel/__freexl.h]: My commit to use conan was not complete: I just renamed freexl.h to __freexk.h, to be sure that this file won't be used. I did not remove it, because I wanted Shun to test first whether the conan version compiles and works also on Windows. The CMakefile.txt, Install.cmake and Libraries.cmake must be changed also.

Comment (JorisGoosen) [reply Desktop/data/importers/excel/__freexl.h]: But I also dont see why this file would be needed without conan?

I know I did something like this with readstat but that was to make it play nice with c++ I think

**Comment by shun2wang** [comment]
Thanks for your reviews! OK let's use conan version to make build easier. and ignore 700 cols limitation. then I will following the official release. 🍻

File: Desktop/data/importers/excel/__freexl.h
Comment (shun2wang) [reply Desktop/data/importers/excel/__freexl.h]: Because when using cmake to reference `freexl.h` in the msys2(Rtools) include path, it will automatically link some other standard libraries, which will conflict with the msvc compiler and linker during compilation. However, once we use conan, we do not need to introduce this header.

**Review by boutinb — APPROVED** [review]
It works really well on Mac.
I think we can merge this for 0.19.2

**Comment by JorisGoosen** [comment]
I agree!

Very nice @shun2wang  😃

**Pull request merged**
